doc/faup.1
.\" Written by Sebastien Tricaud (c) 2012, 2013, 2014
.\" http://www.github.com/stricaud/faup
.Dd August 5,2014
.Dt FAUP 1
.Os
.Sh NAME
.Nm faup
.Nd Finally An Url Parser!
.Sh SYNOPSIS
.Nm faup
.Op Fl ablptu
.Op Fl d Ar delim
.Op Fl f Ar scheme | credential | subdomain | domain | domain_without_tld | host | tld | port | resource_path | query_string | fragment
.Op Fl m Ar module1 module2 ...
.Op Fl o Ar csv | json | module
.Op Fl r Ar count
.Op Fl w Ar ip:port
.Ao Ar url | file Ac
.Sh DESCRIPTION
Everytime 
.Nm faup 
is called with an
.Ar url
or a 
.Ar file 
containing a list of URLs, it will split the various parts of it.
.Pp
.Nm faup
is designed to be really fast to do that operation, much faster than a regex, and give you much more features. The problem is harder than
what it first appears. If you want to get fields like subdomains from an URL, you will need to extract the TLD properly. And since extracting
the TLD properly is impossible with a regex, you will be stuck.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl a
skip provided argument file open check
.It Fl b
Run the webserver in background
.It Fl d Ar delim
Use
.Ar delim
as the field delimiter character instead of the comma character.
.It Fl f Ar field
Extract only the wanted
.Ar field
which can be one of the following: scheme, credential, subdomain, domain, domain_without_tld, host, tld, port, resource_path, query_string, fragment
.It Fl l
Prefix each line with the line number. Works only if the output is CSV.
.It Fl m Ar module1 module2 ..
Load the modules in the wanted order. If the list is empty, 
.Nm faup
will run without any module.
.It Fl o Ar format
Output in the wanted 
.Ar format
which can be one of the following: csv, json, module
.It Fl p
print the header (applies to CSV only)
.It Fl r Ar count
Remove the
.Ar count
number of characters from the end. Useful to remove the dot in DNS urls given.
.It Fl t
do not extract TLD > 1 (eg. only get 'uk' instead of 'co.uk')
.It Fl u
Update the TLD suffix list
.It Fl w Ar ip:port
Start
.Nm faup
in webserver mode

.Sh EXIT STATUS
.Ex -std

.Sh EXAMPLES
Extract the TLD from slashdot.org:
.Pp
.Dl "faup -f tld www.slashdot.org"
.Pp
Run faup in webserver mode in background:
.Pp
.Dl "faup -b -w 0.0.0.0:9876"
.Pp
List all the available modules:
.Pp
.Dl "faup $ modules list all"
.Pp
Enable the uppercase lua module:
.Pp
.Dl "faup $ modules enable uppercase.lua"
.Pp
Output the wanted url in json
.Pp
.Dl "faup -o json http://www.example.co.uk:1234/foo.html#bar"
.Pp

.Sh ENVIRONMENT
The following environment variable affects the exectution of 
.Nm faup
:
.Bl -tag -width ".Ev CLICOLOR_FORCE"
.It Ev FAUP_DATA_DIR
If the environment variable 
.Ev FAUP_DATA_DIR
is set, tell
.Nm faup
where to find its datadir, to get modules or the tld prefix list ('mozilla.tlds')




src/lib/bindings/python/pyfaup/faup.py
import sys

from .functions import *

class UrlNotDecoded(Exception):
    pass

class Faup(object):
    """
    Faup Python Library 
    """
    def __init__(self):
        self.options = faup_options_new()
        self.fh = faup_init(self.options)
        self.decoded = False
        self.retval = {}

    def __del__(self):
        faup_terminate(self.fh)
        faup_options_free(self.options)

    def decode(self, url):
        """
        This function creates a dict of all the url fields.

        :param url: The URL to normalize
        """
        self.url = None

        # Alright, I assume no one is using python 1.x nor 4.x
        if sys.version.split('.')[0].split('.')[0]=='3':
            self.url = bytes(url,'utf-8')
        if sys.version.split('.')[0].split('.')[0]=='2':
            self.url = bytes(url)

        self.url = faup_decode(self.fh, self.url, len(url))
        self.decoded = True
        self.retval = {}
        
    def get_version(self):
        return faup_get_version()

    def _get_param_from_pos_and_size(self, pos, size):
        if pos < 0:
            return None

        else:
            return self.url[pos:(pos+size)]

    def get_scheme(self):
        """
        Get the scheme of the url given in the decode function

        :returns: The URL scheme
        """
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_scheme_pos(self.fh)
        size = faup_get_scheme_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)

    def get_credential(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_credential_pos(self.fh)
        size = faup_get_credential_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)

    def get_subdomain(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")


        pos = faup_get_subdomain_pos(self.fh)
        size = faup_get_subdomain_size(self.fh)
        return self._get_param_from_pos_and_size(pos, size)

    def get_domain(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_domain_pos(self.fh)
        size = faup_get_domain_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)

    def get_domain_without_tld(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_domain_without_tld_pos(self.fh)
        size = faup_get_domain_without_tld_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)
        
    def get_host(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_host_pos(self.fh)
        size = faup_get_host_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)
        
    def get_tld(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_tld_pos(self.fh)
        size = faup_get_tld_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)
            
    def get_port(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_port_pos(self.fh)
        size = faup_get_port_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)

    def get_resource_path(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_resource_path_pos(self.fh)
        size = faup_get_resource_path_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)

    def get_query_string(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_query_string_pos(self.fh)
        size = faup_get_query_string_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)

    def get_fragment(self):
        if not self.decoded:
            raise UrlNotDecoded("You must call faup.decode() first")

        pos = faup_get_fragment_pos(self.fh)
        size = faup_get_fragment_size(self.fh)

        return self._get_param_from_pos_and_size(pos, size)

    def get(self):        
        self.retval["scheme"] = self.get_scheme()
        self.retval["tld"] = self.get_tld()    
        self.retval["domain"] = self.get_domain()
        self.retval["domain_without_tld"] = self.get_domain_without_tld()
        self.retval["subdomain"] = self.get_subdomain()    
        self.retval["host"] = self.get_host()
        self.retval["port"] = self.get_port()
        self.retval["resource_path"] = self.get_resource_path()
        self.retval["query_string"] = self.get_query_string()
        self.retval["fragment"] = self.get_fragment()
        return self.retval
